Clean up obsolete `prometheus-` folder from Prometheus volumes

vicwicker opened the pull request:
**How to categorize this PR?**

/area monitoring
/kind enhancement

**What this PR does / why we need it**:

We have recently discovered that some Prometheus volumes still contain the `prometheus-` folder with old Prometheus data. This folder was used for storage before [GEP-19](https://github.com/gardener/gardener/issues/9065), when the Prometheus operator was introduced. With the operator, the storage folder is now `prometheus-db`:

```bash
> k get pods prometheus-shoot-0 -o json | jq -r '
    (.spec.volumes[] | select(.name == "prometheus-db")),
    (.spec.containers[].volumeMounts[] | select(.name == "prometheus-db"))'
{
  "name": "prometheus-db",
  "persistentVolumeClaim": {
    "claimName": "prometheus-db-prometheus-shoot-0"
  }
}
{
  "mountPath": "/prometheus",
  "name": "prometheus-db",
  "subPath": "prometheus-db"
}
```

In the past, the `prometheus-` subpath was mounted instead. For example, in the [seed](https://github.com/gardener/gardener/blob/96335a6487944f05322ac8f182f46c1e738ee2ff/pkg/component/monitoring/charts/seed-monitoring/charts/core/charts/prometheus/templates/prometheus.yaml#L134-L135).

GEP-19 included a [migration](https://github.com/gardener/gardener/pull/9128/commits/ca0a1863fca24146d6689bdb6b8d9017a44441d9#diff-c6e1ff35f50162ec0f10e92c7b5a3407c6d314b2cde196661c0d5253120d9c71R101-R106) to transfer data from the old folder to the new one. However, it appears there are remnants where the `prometheus-` folder is either a sibling of the new `prometheus-db` or a subfolder. Unfortunately, pinpointing the exact sequence of steps that led to this situation is now very challenging.

In some cases, the old `prometheus-` folder has been observed consuming up to 7GB of data. This poses a risk of Prometheus running out of disk space, as it operates, for example, in a control plane Prometheus, on a 20GB volume with a disk retention setting of 15GB.

This PR introduces an init container to Prometheus to remove the obsolete `prometheus-` folder. However, this init container is injected differently based on the Prometheus type:

- In shoot Prometheus instances, where reconciliation cannot be guaranteed to successfully occur, such init containers are injected during the startup of the `gardenlet`. Hibernated shoot Prometheus instances are accordingly woken up to ensure the cleanup runs. The `gardenlet` patches Prometheus custom resources on-the-fly, bypassing the previously ignored ManagedResource. The GRM reverts these patches once the ManagedResource is unignored upon completion. Finally, an annotation marks those shoot Prometheus instances successfully cleaned up:

```
monitoring.resources.gardener.cloud/prometheus-obsolete-folder-cleaned-up: "true"
```

- Other Prometheus instances include the aforementioned init container as part of their reconciliation.

These cleanup steps will be undone in upcoming releases.

**Special notes for your reviewer**:

/cc @istvanballok @rickardsjp @chrkl @rfranzke 

**Release note**:

```other operator
Clean up obsolete `prometheus-` folder from Prometheus volumes. This might be a leftover of GEP-19
```

vicwicker committed:
Clean up obsolete "prometheus-" folder from Prometheus volumes

In GEP-19, the Prometheus operator was introduced in Gardener to manage
Prometheus instances. This operator configures Prometheus statefulsets
to use the "prometheus-db" folder as a subpath for storage:

    > k get pods prometheus-shoot-0 -o json | jq -r '
        (.spec.volumes[] | select(.name == "prometheus-db")),
        (.spec.containers[].volumeMounts[] | select(.name == "prometheus-db"))'
    {
      "name": "prometheus-db",
      "persistentVolumeClaim": {
        "claimName": "prometheus-db-prometheus-shoot-0"
      }
    }
    {
      "mountPath": "/prometheus",
      "name": "prometheus-db",
      "subPath": "prometheus-db"
    }

Previously, the "prometheus-" folder was used instead, as in:

    https://github.com/gardener/gardener/blob/96335a6487944f05322ac8f182f46c1e738ee2ff/pkg/component/monitoring/charts/seed-monitoring/charts/core/charts/prometheus/templates/prometheus.yaml#L134-L135

During the migration to the Prometheus operator, there was a disk
migration to move data between the two folders:

    https://github.com/gardener/gardener/pull/9128/commits/ca0a1863fca24146d6689bdb6b8d9017a44441d9#diff-c6e1ff35f50162ec0f1[…]1c0d5253120d9c71R101-R106

However, the old "prometheus-" folder still exists in some volumes,
consuming up to e.g., 7GB of data in certain control-plane Prometheus
instances. This poses a risk of Prometheus running out of disk space, as
it operates on a 20GB volume with a disk retention setting of 15GB. In
some cases, we have seen the old "prometheus-" folder as a sibling of
"prometheus-db". In some other cases, as a child.

We fix this leftover by introducing a temporal init container in
Prometheus to remove the obsolete folders.

Unfortunately, identifying the exact sequence of steps
or potential disk migration failures that led to this situation is
challenging after this time and will remain unresolved.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

istvanballok reviewed (commented):
/lgtm

rfranzke commented on pkg/component/observability/monitoring/prometheus/prometheus.go:
```diff
@@ -83,6 +83,17 @@ func (p *prometheus) prometheus(cortexConfigMap *corev1.ConfigMap) *monitoringv1
+						Name:      "prometheus-db",
+						MountPath: "/prometheus",
+					}},
+				}},
```
We cannot assume that all `Prometheus` of all `Shoot`s have been reconciled within 3 releases. For them, I guess we have to add this `init` container in `app/gardenlet/migration.go`. Only then we can be sure that this code was executed before we remove it after 3 releases.

vicwicker replied:
I addressed your suggestions and accordingly adapted `cmd/gardenlet/app/migration.go`. I reworked the PR description and tried to capture a general explanation of how this cleanup now works. You can also find the relevant explanations below, and more details on each commit message. 

> GEP-19 included a [migration](https://github.com/gardener/gardener/pull/9128/commits/ca0a1863fca24146d6689bdb6b8d9017a44441d9#diff-c6e1ff35f50162ec0f10e92c7b5a3407c6d314b2cde196661c0d5253120d9c71R101-R106) to transfer data from the old folder to the new one. However, it appears there are remnants where the `prometheus-` folder is either a sibling of the new `prometheus-db` or a subfolder. Unfortunately, pinpointing the exact sequence of steps that led to this situation is now very challenging.
> 
> In some cases, the old `prometheus-` folder has been observed consuming up to 7GB of data. This poses a risk of Prometheus running out of disk space, as it operates, for example, in a control plane Prometheus, on a 20GB volume with a disk retention setting of 15GB.
>
> This PR introduces an init container to Prometheus to remove the obsolete `prometheus-` folder. However, this init container is injected differently based on the Prometheus type:
> 
> - In shoot Prometheus instances, where reconciliation cannot be guaranteed to successfully occur, such init containers are injected during the startup of the `gardenlet`. Hibernated shoot Prometheus instances are accordingly woken up to ensure the cleanup runs. The `gardenlet` patches Prometheus custom resources on-the-fly, bypassing the previously ignored `ManagedResource`. The GRM reverts these patches once the `ManagedResource` is unignored upon completion. Finally, an annotation marks those shoot Prometheus instances successfully cleaned up:
> ```
> monitoring.resources.gardener.cloud/prometheus-obsolete-folder-cleaned-up: "true"
> ```
> - Other Prometheus instances include the aforementioned init container as part of their reconciliation.
> 
> These cleanup steps will be undone in upcoming releases.

vicwicker committed:
Prepare migration to remove shoot Prometheus obsolete folders

This commit is only the skeleton to clean up shoot Prometheus folders as
part of the startup of the gardenlet. It iterates over the Cluster
resources in the seed to find shoots, but executes nothing yet.

vicwicker committed:
Patch shoot Prometheus number of replicas to 1

This guarantees the init container runs for hibernated shoots. For
running shoots, setting the number of replicas to 1 is a no-op.

vicwicker committed:
Ignore shoot Prometheus ManagedResource before patching

Shoot Prometheus instances are actually managed by the GRM. This means
any patches directly applied to the Prometheus custom resources will be
instantly reverted unless the corresponding ManagedResource is ignored
via annotation:

    resources.gardener.cloud/ignore="true"

Upon completion, the ManagedResource can be unignored for the GRM to
reset the Prometheus state. This is done in a deferred call to make sure
the annotation is always removed, also in case of failures.

If the gardenlet process is killed before the cleanup tasks are
complete, the ignore annotation could be leaked. To guarantee the state
is eventually consistent, the gardenlet makes sure to remove ignore
annotations before start up.

vicwicker committed:
Add annotation to avoid reiterating attempts to clean up

This commit uses a temporal annotation to mark shoot Prometheus
instances already cleaned up. This way, the gardenlet doesn't repeatedly
attempt to clean up every shoot Prometheus on every restart.

This temporal annotation will be removed in a second migration step.

vicwicker committed:
Allow the gardenlet to read Prometheus resources

This is necessary for the gardenlet to be able to read and patch
Prometheus resources.

vicwicker committed:
Wait for shoot Prometheus pod to run

This commit prepares the shoot Prometheus cleanup by creating an empty
patch for Prometheus resources, applying it and waiting for shoot
Prometheus pods to be in running state.

The following commits will populate such empty patch.

vicwicker committed:
Patch shoot Prometheus to inject init container

This commit makes sure all shoot Prometheus instances receive the init
container that removes the obsolete "prometheus-" folders. Such init
container is only injected if it is still not present in the Prometheus
resource.

vicwicker committed:
Do not add the init container for shoot Prometheus instances

Shoot Prometheus instances have the additional challenge of not being
guaranteed successful reconciliation. For instance, if shoots are
hibernated or if the reconciliation fails.

In consequence, the previous commit is now tailored specifically for
Prometheus instances in the garden and seed clusters, where
reconciliation is expected to work. A more advanced migration for
shoot Prometheus instances will be addressed in the subsequent commits.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Execute shoot Prometheus clean up asynchronously

If there are many hibernated shoots that need to be woken up, new nodes
will need to be spun up, delaying the startup of the gardenlet in few
minutes.

This commit improves this by letting this cleanup run in the background.
Note this is implemented in a way that it can never return an error.
Failures can still happen and errors are properly logged but they are
not propagated back to the gardenlet.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Mark new shoot Prometheus instances as cleaned up

This way, the gardenlet startup does not execute the cleanup for new
shoot Prometheus instances.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Do not reset cleanup annotation during reconciliation

We need to preserve when a shoot Prometheus resource is marked as
cleaned up so that gardenlet restarts do not keep attempting to clean up
Prometheus instances that were already done.

When a new shoot Prometheus instance is deployed, the GRM detects the
Prometheus resource does not exist yet and creates it directly with the
cleaned up mark. However, the first time the shoot is reconciled the
Prometheus resource exists and the mark is reverted. This is not the
case for migrated shoot Prometheus instances, where the annotation is
owned by the gardenlet.

To fix this, this commit makes sure the GRM preserves the annotation
across shoot reconciliations.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

rfranzke commented on cmd/gardenlet/app/migration.go:
```diff
@@ -237,3 +245,161 @@ func cleanupNginxConfigMaps(ctx context.Context, client client.Client) error {
+			}
+
+			if !hasInitContainer {
+				return retry.MinorError(fmt.Errorf("prometheus-shoot-0 pod does not have the cleanup-obsolete-folder init container"))
```
Maybe also log something here?

rfranzke commented on cmd/gardenlet/app/migration.go:
```diff
@@ -237,3 +245,161 @@ func cleanupNginxConfigMaps(ctx context.Context, client client.Client) error {
+	log.Info("Clean up Prometheus obsolete folders")
+
+	clusterList := &extensionsv1alpha1.ClusterList{}
+	if err := seedClient.List(ctx, clusterList); err != nil {
```
You should ignore shoots in deletion or with a deleting namespace

rfranzke commented on cmd/gardenlet/app/migration.go:
```diff
@@ -237,3 +245,161 @@ func cleanupNginxConfigMaps(ctx context.Context, client client.Client) error {
+			managedResource.Annotations[resourcesv1alpha1.Ignore] = "true"
+			if err := seedClient.Patch(ctx, managedResource, managedResourcePatch); err != nil {
+				log.Error(err, "Failed to ignore ManagedResource for Prometheus", "cluster", cluster.Name)
+				return nil
```
Why doesn't this return the error? Nobody will notice it if the gardenlet just silently continues. Same above and below

vicwicker replied:
Yeah, I guess that's a bit unorthodox but that is on purpose. Let me explain :)

I don't see this clean up as a hard requirement for the `gardenlet` to start but rather as a best-effort attempt to clean up some leftovers from the past. AFAIK, if this fails and we return the error, the `gardenlet` won't start and the update will be blocked for what I would consider no good reason. Furthermore, returning the error means the `gardenlet` needs to wait for all control plane Prometheis to be cleaned up on the first start, which could be an expensive operation because it involves restarting the pod and wait for it to get to running state. This is specially heavy for hibernated shoots, where the shoot Prometheus stateful set needs to be additionally scaled. If there are many hibernated shoots, the likelihood of requiring a new node to have capacity for all those now temporarily awake Prometheus stateful sets increases and, with that, the `gardenlet` start up could get delayed by minutes.

With the current approach where errors are ignored and the clean up runs asynchronously, we risk that errors are not noticed and some leftovers not cleaned up. However, I would expect most occurrences are successfully cleaned up, which makes it better, although not perfect.

Anyhow, these were my reasons to go this way instead of returning the error. Nevertheless, as I said above, I understand your concern that this might look a bit strange. If you think those reasons are not good enough or in general you prefer returning the error, then I'm also fine changing that and making this clean up a "must-run" for the `gardenlet`.

rfranzke replied:
If we don't care if the cleanup succeeds, why do we do it in the first place then? :) I understand where you're coming from, but still, even if it takes a few minutes - if it is important to us to perform this cleanup (and apparently it is - otherwise, we wouldn't have spent time with this PR), then that's what it takes. Ideally, this is a one-time action, and even if it fails for some shoots and needs to be retried after a gardenlet restart, it will be faster (because it probably has succeeded for some other shoots, i.e., the total number of to-be-migrated-shoots is smaller).

TL;DR: I would definitely return the error and do it properly - otherwise, failures might remain unnoticed, and we can never be sure whether we successfully performed the cleanup everywhere.

vicwicker replied:
We do this because there is an undesired additional disk usage in those Prometheus volumes with an obsolete folder. It is not so harmful, except for the low risk of causing a Prometheus instance to run out of disk - which we have seen only once so far.

I see your point about the value of making sure all Prometheus instances are cleaned up, and accepting the possible overhead on the `gardenlet`.

Alright then, I will change this.

rfranzke commented on cmd/gardenlet/app/migration.go:
```diff
@@ -237,3 +245,161 @@ func cleanupNginxConfigMaps(ctx context.Context, client client.Client) error {
+		unignoreManagedResource(ctx, log, cluster.Name)
+
+		tasks = append(tasks, func(ctx context.Context) error {
+			managedResource, managedResourcePatch, err := getManagedResourceWithPatch(ctx, cluster.Name)
```
What if the managed resource is not found/does not exist? This should be tolerated, I guess

rfranzke commented on cmd/gardenlet/app/migration.go:
```diff
@@ -237,3 +245,161 @@ func cleanupNginxConfigMaps(ctx context.Context, client client.Client) error {
+
+	go func() {
+		_ = flow.Parallel(tasks...)(ctx)
+	}()
```
Return the error...?

vicwicker replied:
I replied [above](https://github.com/gardener/gardener/pull/12219#discussion_r2166574012).

vicwicker committed:
Review: Return errors

Ignoring the errors will guarantee cleaning up the obsolete
"prometheus-" folders won't interfere with the startup of the gardenlet.
However, it won't guarantee that all Prometheus volumes will be
successfully cleaned up once the migration code is removed.

With this commit, we favor ensuring the obsolete folders are correctly
cleaned up and accept the additional overhead in the gardenlet startup.
If an error occurs, now the gardenlet will fail and restart.

vicwicker committed:
Review: Improve logging while waiting for the pod to be ready

The retry method does not log returned errors between retries. This
commit improves this by logging the error before returning.

vicwicker committed:
Refactor Error Handling for Unignoring ManagedResource

Ensure proper error handling when unignoring a ManagedResource:

- Log potential unignore errors but return the original error if cleanup
  fails.
- Return the unignore error if cleanup succeeds but unignore fails.

This update removes the defer call to unignore the ManagedResource,
allowing explicit logging or returning of unignore errors as needed. To
enhance code readability and minimize error checks, the unignore
operation is now performed as late as possible.

vicwicker committed:
Review: Skip shoots being deleted

vicwicker committed:
Review: Check if the Cluster resource is available in the seed

After returning errors in previous commit, a new error might be returned
if the Cluster resource was not yet installed in the seed when the
gardenlet runs.

rfranzke reviewed (commented):
/lgtm
/approve
